Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar (CFG), which is a grammar that has a
+single nonterminal symbol on the left-hand side of each rule.
+Derivation by a grammar is inherently nondeterministic.
+-/
+
+/--
+The type of symbols is the disjoint union of terminals `T` and nonterminals `N`.
+-/
+inductive Symbol (T : Type _) (N : Type _)
+  /-- Terminal symbols (of the same type as the language) -/
+  | terminal    (t : T) : Symbol T N
+  /-- Nonterminal symbols (must not be present at the end of word being generated) -/
+  | nonterminal (n : N) : Symbol T N
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type _) where
+  /-- Type of nonterminals -/
+  NT : Type _
+  /-- Initial nonterminal -/
+  initial : NT
+  /-- Rewrite rules -/
+  rules : List (NT × List (Symbol T NT))
+
+namespace ContextFreeGrammar
+
+variable {T : Type _}
+
+/-- One step of context-free transformation. -/
+def Produces (g : ContextFreeGrammar T) (w₁ w₂ : List (Symbol T g.NT)) : Prop :=
+  ∃ r ∈ g.rules,
+    ∃ u v : List (Symbol T g.NT),
+      w₁ = u ++ [Symbol.nonterminal r.fst] ++ v ∧ w₂ = u ++ r.snd ++ v
```

Comment:
This subexpression feels suboptimal to me; it seems annoying both to prove and to use. Some things you might want to consider:
* extracting this to its own `abbrev RewritesTo (rule : NT × List (Symbol T NT)) (w₁ w₂ : List (Symbol T g.NT))`, so that you have a namespace to add API to
* defining a custom `inductive` relation
* using a natural number index

I think these last two have the property that they should be straightforward to write a `DecidableRel` instance for, which will mean that you give the `dec_trivial` tactic a major power up for context-free grammars.

Reply:
Here's what the inductive type would look like:
```lean
inductive RewritesTo {T NT : Type _} (rule : NT × List (Symbol T NT)) :
    List (Symbol T NT) -> List (Symbol T NT)  -> Prop
  /-- The replacement is at the start of the remaining string. -/
  | head (xs : List (Symbol T NT)) : RewritesTo rule (.nonterminal rule.1 :: xs) (rule.2 ++ xs)
  /-- There is a replacement later in the string. -/
  | cons (x : Symbol T NT) {xs₁ xs₂ : List (Symbol T NT)} (h : RewritesTo rule xs₁ xs₂) : RewritesTo rule (x :: xs₁) (x :: xs₂)
```

Reply:
I don't understand your first suggestion. What should it do?

Reply:
Does it mean that a sequence of rewriting could be evaluated by Lean, as if a deterministic function on the strings?
(always the first ocurence of rewrite-able nonterminal would be rewritten?)

Reply:
The first suggestion is to simply pull out a definition
```lean4
abbrev RewritesTo (rule : NT × List (Symbol T NT)) (w₁ w₂ : List (Symbol T g.NT)) : Prop :=
  ∃ u v : List (Symbol T g.NT),
      w₁ = u ++ [Symbol.nonterminal r.fst] ++ v ∧ w₂ = u ++ r.snd ++ v
```
then
```suggestion
    r.RewritesTo w₁ w₂
```

Reply:
My second question was about the dec_trivial thing, i.e., the motivation for the change.

Reply:
The dec_trivial thing would mean that for any concrete language, `dec_trivial` can automatically prove whether there is a *single* rewrite between two strings.

Reply:
But the other motivation is that I think it will be a lot easier to prove things using the inductive definition.

Reply:
I think I get it now. Thanks!